proposals/Imperative-Shadow-DOM-Distribution-API.md

```diff
@@ -1,48 +1,70 @@
 
-The *manual* means "we support only imperative APIs for the shadow tree".
-The default is "we support only declarative API for the shadow tree".
+[assigned nodes]: https://dom.spec.whatwg.org/#slot-assigned-nodes
 
+An `HTMLSlotElement` gets a new API, called `assign()` (tentative name), and a new property, `manually-assigned-nodes`,
+which is an ordered list of nodes. Unless stated otherwise, it is empty.
 
-# Imperative Slotting API
+Basically, `slot.assign(sequence<Node> nodes)` sets the slot's `manually-assigned-nodes` to `nodes`. See the later section for details.
 
-In addition to [assigned nodes], which is already defined in DOM Standard,
-a slot has an associated *manually-assigned-nodes* (ordered list). Unless stated otherwise, it is empty.
+- *manually-assigned-nodes* is an internal field. It is write-only. Users cannot read the value directly.
 
-[assigned nodes]: https://dom.spec.whatwg.org/#slot-assigned-nodes
+- *manually-assigned-nodes* are different than [assigned nodes]. They are candidates for slot assignments. However, we check each node's slotable validity. If an invalid node is detected, we throw an exception in `slot.assign()` and clear `manually-assigned-nodes`. If the validity check is successful, the browser recalculates [assigned nodes] later, based on `manually-assigned-nodes`. The calculated [assigned nodes] are then observable.
 
-A slot gets new API, called *assign* (tentative name).
+See examples below.
 
-Basically, `slot.assign(sequence<Node> nodes)` sets the slot's *manually-assigned-nodes* to *nodes*. See the later section for details.
+## Declarative API and Imperative API should be exclusively used in each shadow tree
 
-- *manually-assigned-nodes* is an internal field. It is write-only. Users cannot read the value directly.
+Mixing declarative API and imperative API would be troublesome and can cause confusion for web developers.
+We *can* invent complex rules. However, no one wants to remember complex rules. Also, supporting both in the same shadow tree would make a browser engine complex.
 
-- *manually-assigned-nodes* are different than [assigned nodes]. For example, users can pass a node which is not host's children. We don't throw an exception even if such a *invalid* node is passed, however, *invalid* node is never selected as [assigned nodes]; the engine recaluculate [assigned nodes] later, based on *manually-assigned-nodes*. Invalid nodes in *manually-assigned-nodes* are simply ignored. The caculataed [assigned nodes] are only observable.
+Thus, we don't allow mixing the declarative and imperative APIs in the same shadow tree. Web developers must explicitly opt-in
+to use an imperative API for each shadow tree. They do this with a new "slotAssignment" parameter on attachShadow():
 
-See also the Example 3 later.
+```js
+const sr = attachShadow({ mode: ..., (optional) slotAssignment: 'manual' });
```

Comment:
This isn't valid JavaScript syntax, so it's kind of confusing. I suggest:

```suggestion
const sr = attachShadow({ mode: 'open', slotAssignment: 'manual' });
```

Reply:
Update